src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,686 @@
+
+
+-- suc2*∘suc≗suc∘2suc :  suc2* ∘ suc ≗ suc ∘ 2suc   -- is by  const refl
+
+suc2*-as∘ :  suc2* ≗ suc ∘ 2*
+suc2*-as∘ 0#        = refl
+suc2*-as∘ (2suc x)  = refl
+suc2*-as∘ (suc2* x) =
+  begin suc2* (suc2* x)     ≡⟨ cong suc2* (suc2*-as∘ x) ⟩
+        suc2* (suc 2x)      ≡⟨⟩
+        suc (2suc 2x)       ≡⟨ cong suc (2suc-as∘ 2x) ⟩
+        suc (2* (suc 2x))   ≡⟨ cong (suc ∘ 2*) (sym (suc2*-as∘ x)) ⟩
+        suc (2* (suc2* x))
+  ∎
+  where 2x = 2* x
+
+---------------------------
+pred∘suc :  pred ∘ suc ≗ id
+pred∘suc 0#        =  refl
+pred∘suc (2suc x)  =  sym (2suc-as∘ x)
+pred∘suc (suc2* x) =  refl
+
+
+suc∘pred :  (x : Bin) → x ≢ 0# → suc (pred x) ≡ x
+
+suc∘pred 0#        0≢0 =  contradiction refl 0≢0
+suc∘pred (2suc _)  _   =  refl
+suc∘pred (suc2* x) _   =  begin suc (pred (suc2* x))   ≡⟨⟩
+                                suc (2* x)             ≡⟨ sym (suc2*-as∘ x) ⟩
+                                suc2* x
+                          ∎
+
+------------------------------
+0+ :  (x : Bin) → 0# + x ≡ x
+0+ _ = refl
+
++0 :  (x : Bin) → x + 0# ≡ x
++0 0#        = refl
++0 (2suc _)  = refl
++0 (suc2* _) = refl
+
+---------------
+2^ :  ℕ → Bin
+2^ 0      =  1#
+2^ (1+ n) =  2* (2^ n)
+
+
+------------
+infixl 7 _*_
+
+_*_ : Op₂ Bin
+
+0#        * _         =  0#
+_         * 0#        =  0#
+(2suc x)  * (2suc y)  =  2* (2suc (x + (y + x * y)))
+(2suc x)  * (suc2* y) =  2suc (x + y * (2suc x))
+                      --
+                      -- 2(1+x) * (1+2y) =  2(1 + 2y + x + 2xy)
+                      --                 =  2(1 + x + y*2(1 + x))
+
+(suc2* x) * (2suc y)  =  2suc (y + x * (2suc y))
+(suc2* x) * (suc2* y) =  suc2* (x + y * (suc2* x))
+            --
+            -- (1 + 2x)(1 + 2y) =  1 + (2y + 2x + 4xy)
+            --                     1 + 2(x + y * (1 + 2x))
+
+------------
+infixl 8 _^_
+
+_^_ : Bin → ℕ → Bin
+_ ^ 0      =  1#
+x ^ (1+ n) =  x * (x ^ n)
+
+------------------------------
+0* :  (x : Bin) → 0# * x ≡ 0#
+0* _ =  refl
+
+*0 :  (x : Bin) → x * 0#  ≡ 0#
+*0 0#        =  refl
+*0 (2suc _)  =  refl
+*0 (suc2* _) =  refl
+
+
+x*y≡0⇒⊎ :  {x y : Bin} → x * y ≡ 0# → x ≡ 0# ⊎ y ≡ 0#
+
+x*y≡0⇒⊎ {0#}      {_ }      _ =  inj₁ refl
+x*y≡0⇒⊎ {_}       {0#}      _ =  inj₂ refl
+x*y≡0⇒⊎ {2suc _}  {2suc _}  ()
+x*y≡0⇒⊎ {2suc _}  {suc2* _} ()
+x*y≡0⇒⊎ {suc2* _} {2suc _}  ()
+x*y≡0⇒⊎ {suc2* _} {suc2* _} ()
+
+nz*nz :  {x y : Bin} → x ≢ 0# → y ≢ 0# → x * y ≢ 0#
+nz*nz x≢0 y≢0 xy≡0  with x*y≡0⇒⊎ xy≡0
+...                 | inj₁ x≡0 =  x≢0 x≡0
+...                 | inj₂ y≡0 =  y≢0 y≡0
+
+
+setoid : Setoid 0ℓ 0ℓ
+setoid = PE.setoid Bin
+
+2*≢1 :  {x : Bin} → 2* x ≢ 1#
+2*≢1 {0#}      ()
+2*≢1 {2suc _}  ()
+2*≢1 {suc2* _} ()
+
+------------------------------------------------------------------------------
+suc≗1+ :  suc ≗ (1# +_)
+suc≗1+ 0#        =  refl
+suc≗1+ (2suc _)  =  refl
+suc≗1+ (suc2* _) =  refl
+
+------------------------------------------------------
+suc+assoc :  (x y : Bin) → (suc x) + y ≡ suc (x + y)
```

Comment:
Again please check the name for this property in Data.Nat.Properties

Reply:
Can you, please, explain?

For ``ℕ,   (suc m) + n ≡ suc (m + n)``  is proved by ``refl``. Also there is ``+-assoc`` for ``ℕ``.

But in Bin0,  ``suc ≗ (1# +_)``  is proved in three patterns, so that
``(suc m) + n ≡ suc (m + n)``  cannot be directly expressed as  +-assoc.
-- ?

Reply:
In `Data.Integer.Properties` it's called [`suc-+`](https://agda.github.io/agda-stdlib/Data.Integer.Properties.html#28581) for instance.